src/vos/vos_common.c

```diff
@@ -243,8 +286,26 @@ vos_tx_end(struct vos_container *cont, struct dtx_handle *dth_in,
 		D_INIT_LIST_HEAD(&tmp.dth_deferred_nvme);
 	}
 
+	if (dth->dth_local) {
+		pool = vos_hdl2pool(dth_in->dth_poh);
+	} else {
+		pool = cont->vc_pool;
+	}
+
 	if (rsrvd_scmp != NULL) {
 		D_ASSERT(nvme_exts != NULL);
+		if (dth->dth_rsrvd_cnt > 0 && dth->dth_rsrvd_cnt >= dth->dth_modification_cnt) {
+			/*
+			 * Just do your best to release the SCM reservation. Can't handle another
+			 * error while handling one already anyway.
+			 */
+			(void)vos_publish_scm(vos_pool2umm(pool), *rsrvd_scmp, false /* publish */);
+			D_FREE(*rsrvd_scmp);
+			*rsrvd_scmp = NULL;
+			err         = -DER_NOMEM;
+			goto cancel;
```

Comment:
There is a silent assumption that the allocation passed as `rsrvd_scmp` is basically taken ownership of and is passed to the `dth->dth_rsrvds[]`. So later on e.g. `vos_ioc_reserve_fini()` assumes `umem_rsrvd_act_cnt(ioc->ic_rsrvd_scm) == 0`. Since in this case the `dth->dth_rsrvds[]` can't accept another record it is disposed of and the whole transaction is aborted.

It is necessary since when doing transactions locally it is sometimes tricky to correctly calculate the number of submodifications upfront. Hence, it is good to have a safeguard in this case.

Ref: https://daosio.atlassian.net/browse/DAOS-11404?focusedCommentId=123025